python/ray/data/_internal/execution/streaming_executor_state.py

```diff
@@ -257,18 +307,39 @@ def _execution_allowed(
     Returns:
         Whether the op is allowed to run.
     """
+    assert isinstance(global_usage, TopologyResourceUsage), global_usage
     # To avoid starvation problems when dealing with fractional resource types,
     # convert all quantities to integer (0 or 1) for deciding admissibility. This
     # allows operators with non-integral requests to slightly overshoot the limit.
     global_floored = ExecutionResources(
-        cpu=math.floor(global_usage.cpu or 0),
-        gpu=math.floor(global_usage.gpu or 0),
-        object_store_memory=global_usage.object_store_memory,
+        cpu=math.floor(global_usage.overall.cpu or 0),
+        gpu=math.floor(global_usage.overall.gpu or 0),
+        object_store_memory=global_usage.overall.object_store_memory,
     )
     inc = op.incremental_resource_usage()
     inc_indicator = ExecutionResources(
         cpu=1 if inc.cpu else 0,
         gpu=1 if inc.gpu else 0,
         object_store_memory=1 if inc.object_store_memory else 0,
     )
-    return global_floored.add(inc_indicator).satisfies_limit(global_limits)
+
+    # Under global limits; always allow.
+    new_usage = global_floored.add(inc_indicator)
+    if new_usage.satisfies_limit(global_limits):
```

Comment:
Nit: If we tweaked the limit scaling to only apply for object store memory (I don't think we'd want it to apply for CPU/GPU usage for now), is there any reason that we couldn't override `new_usage.object_store_memory` with the downstream memory here and do the single check?
```python
new_usage = global_floored.add(inc_indicator)
downstream_usage = global_usage.downstream_memory_usage[op]
scaled_limit = global_limits.scale_memory(downstream_usage.topology_fraction)
new_usage.object_store_memory = downstream_usage.object_story_memory
return new_usage.satisfies_limit(scaled_limit)
```

Reply:
I think you'd still want to allow the operator to burst above its fraction in the case of normal operation. For example, if the last operator was the memory hungry one, it should be allowed to use almost 100% of the memory allocation for the entire pipeline, if no operators are using memory, even though its "fair share" is 1/N.